docs: klip-33 key format (#6017)
confluentinc/ksql: design-proposals/klip-33-key-format.md

disparate systems, even though it is limited to changing the _values_ format and structure.
Supporting other key formats opens this up to also transforming the key into a different format. 

In some cases users are able to work around this limitation. This may involve changing upstream code,
or introducing pre-processing, or, in the case of Connect, using SMTs to convert the key format. All
such solutions tend to increase the complexity of the system, and generally hurt performance.

To open ksqlDB up to new problems spaces and to drive adoption, ksqlDB should support other key formats. 

## What is in scope

 * Addition of a new optional `KEY_FORMAT` property in the `WITH` clause, to set the key format.
 * Addition of a new optional `FORMAT` property in the `WITH` clause, to set both the key & value formats.
 * Addition of new server configuration to provide defaults for key and value formats.
 * Removal of requirement for `VALUE_FORMAT` property in the `WITH` clause of CT/CS statements, where
   the server configuration provides a default.
 * Support for additional key column data types, where the key format supports it:
    * `DECIMAL`
    * `BOOLEAN`

>  * Support of the following key formats:

Reviewer: I suspect `KAFKA` will be supported as well?

Author: It's already supported. So... adding support is not in scope.

Reviewer 2: Well, but `KEY_FORMAT='KAFKA'` should be a valid option thus it seem it should be listed? It's not about adding support for the format, but about the language change to explicitly set the format.

    * `DELIMITED`: single key columns as a single string value.

>     * `JSON`: single key column as an anonymous value, i.e. not within a JSON object.

Reviewer: out of interest, is this actually any different than `KAFKA`?

Author: Yes, it is different. JSON is serialized as a JSON payload in a string. So, for example where as a `KAFKA` `INT` would be serialized as 4 bytes, a JSON `INT` would be essentially serialized as the `val.toString().getBytes(UTF8);`.

>     * `JSON`: single key column as an anonymous value, i.e. not within a JSON object.

Author: Good catch. Added.

>     * `JSON` / `JSON_SR`: single key column as an anonymous value, i.e. not within a JSON object.

Reviewer 2: Why do we have two values? And what is JSON_SR? (I guess this is just a question for my own education).

Author: ... it's not nice :(

`JSON` format can _read_ a schema from the SR, but doesn't _write_ to the SR.
`JSON_SR` format both _reads_ and _writes_ schemas to SR.

And internally they use different code, so have different behaviour, e.g. when serializing maps (https://github.com/confluentinc/ksql/issues/6049).

I'm not sure of the _why_. @agavra added it, so I'm sure there was good reason.  He can likely comment more.

Reviewer: yeah it was a product decision how we wanted to handle it - the difference is that they are actually different serialization formats and are not byte compatible (one adds the magic byte and the schema header, the other is vanilla json). We could've hacked around this and magically removed the header if it exists, but we decided it would be more confusing to the user than just having two formats

drives me bananas too 🍌

Reviewer 2: Not sure if I understand. For `JSON` if the schema is received from the SR, the must still be some "header" encoding the schema ID to fetch the schema? So how can one for both formats _not_ have a "header"?

Reviewer: `JSON` does not support schema registry integration

Reviewer: >  So how can one for both formats not have a "header"?

>     * `AVRO`: single key column as an anonymous value, i.e. not within an Avro record.

Reviewer 2: If we have structured keys in the future, how can a user set a non-primitive single-column AVRO key format? -- Would it be better to reserve `AVRO` for this case? and use `PRIMITIVE_AVRO` (or similar) for the case you describe? (Similar for JSON actually).

Author: We already support `WRAP_SINGLE_VALUE` in the `WITH` clause to handle switching between anonymous and wrapped serialized forms.   The future work to support wrapped keys will introduce `WRAP_SINGLE_KEY`. 

I've added details to the KLIP.

Author: See https://github.com/confluentinc/ksql/pull/6017/files#diff-f01d9cdc48d983b2fc1e805fbac19028R287-R324

Reviewer 2: Thanks. Seems I missed the memo. `WRAP_SINGLE_VALUE` sounds a little awkward to me personally...

  * Support for reading & writing key schemas to & from the schema registry. 

## What is not in scope

 * Support for multiple key columns: this will come later.
 * Support for single key columns _wrapped_ in an envelope of some kind: this will come later.
 * Support for complex key column data types, i.e. array, struct and map 
 * Support for `PROTOBUF` keys, as this requires support for wrapped keys: this will come later.

## Value/Return

We know from customers and community members that there are a lot of people that have data with 
non-`KAFKA` formatted keys. This is the first step to unlocking that data and use-cases.

With support for `AVRO` and `JSON` key formats there are a lot of existing use-cases that suddenly
no longer require pre-processing, or tricky Connect SMTs configured and there are new use-cases,
which ksqlDB was previously unsuitable for, as documented in the motivation section, which it can 
now handle.

## Public APIS

### CREATE properties

The following new properties will be accepted in the `WITH` clause of `CREATE` statements for streams
and tables.

* `KEY_FORMAT`: sets the key format, works long the same lines as the existing `VALUE_FORMAT`.

> * `FORMAT`: sets both the key and value format with a single property.

Reviewer 2: Do we want to allow to mix KEY_FORMAT, VALUE_FORMAT and FORMAT or not? Ie, we can either say: (1) use only KEY_FORMAT + VALUE_FORMAT or only FORMAT (but don't mix both), or (2) if you mix FORMAT with KEY_FORMAT (or VALUE_FORMAT) the KEY_FORMAT overwrites whatever FORMAT specifies.

Author: As the KLIP states: 

> Providing `FORMAT` along with either `KEY_FORMAT` or `VALUE_FORMAT` will result in an error.

:)

The existing `VALUE_FORMAT` property will no longer be required, if the server config provides a 
default. (See below).

### Server configs

The following new configuration options will be added. These configurations can be set globally, 
within the application property file, or locally, via the `SET` command.

> * `ksql.persistence.default.format.key`: the default key format.

Reviewer 2: Should there also be `ksql.persistence.default.format` ?

Author: I don't think so.  If you're setting config, easy enough to just set both.

* `ksql.persistence.default.format.value`: the default value format.

## Design

### New CREATE properties

The new `KEY_FORMAT` or `FORMAT` property will be supported where ever the current `VALUE_FORMAT` is
supported. Namely:

 * In `CRREATE STREAM` and `CREATE TABLE` statements.

>  * In `CREATE STREAM AS SELECT` and `CREATE TABLE AS SELECT` statements.

Reviewer: I'm worried this might change partitioning in an unexpected way - are there any complications here that we need to think about, or is it always OK because it's a new query anyway?

Author: I can't think of any issues... but that's why we put these designs up for review. If you can... then let me know :D

 
Providing `FORMAT` will set both the key and value formats. Providing `FORMAT` along with either 

> `KEY_FORMAT` or `VALUE_FORMAT` will result in an error.

Reviewer 2: Ah. Here we go -- this should be added above already :)

The key format will follow the same inheritance rules as the current value format. Namely: any 
derived stream will inherit the format of its leftmost source, unless the format is explicitly set
in the with clause.

For example:

```sql
-- Creates a table over a changelog topic with AVRO key and JSON value:
CREATE TABLE USERS (
    ID BIGINT PRIMARY KEY,
    NAME STRING
  ) WITH (
    KAFKA_TOPIC='USERS',
    KEY_FORMAT='AVRO',
    VALUE_FORMAT='JSON'
  );

-- Creates a stream over a topic with JSON key and value:
CREATE STREAM BIDS (
    ITEM_ID BIGINT KEY,
    USER_ID BIGINT,
    AMMOUNT INT
  ) WITH (
    KAFKA_TOPIC='USERS',
    FORMAT='JSON'
  );

> -- Change the key format of a stream:

Reviewer 2: When does the format change happen? Or this this unspecified by definition?

For example, the KS program could be `jsonStreamBids.selectKey(/*change the object from JSON to AVRO*)` (on read) or it could be `selectKey(/*change the object from JSON to AVRO*).to("AVRP_BIDS")` (on write).

For the simple `SELECT *` query from the example there is no difference, but assume you do anything stateful the data format of the state stores might differ -- also if you do joins, it might differ (and might require repartitioning).

Overall, it might be good to give the optimizer some freedom when to do the format change and thus it might be good to only guarantee that the _output_ stream will have the format but no guarantee about intermediate formats.

Author: Internal topics are... internal :).  

I'll add a section to the compatibility section to cover this.

Author: Add https://github.com/confluentinc/ksql/blob/dcd28ff1b26705625a477db4e3e81a1891b184aa/design-proposals/klip-33-key-format.md#internal-topics

In general, I'm saying we don't need to make any guarantees from release to release about the internal formats of topics. We're free to change them as we want.  

Of course, we _do_ maintain the internal formats for _existing_ queries. But that's handled already by capturing the internal format in the query plan.

CREATE STREAM AVRP_BIDS 
  WITH WITH (
    KEY_FORMAT='AVRO'
  ) AS
    SELECT * FROM BIDS;

-- Creates an enriched stream. The key format is inherited from the leftmost source, i.e. JSON:
CREATE STREAM ENRICHED_BIDS AS 
  SELECT *
  FROM BIDS 
   JOIN USERS ON BIDS.USER_ID = USERS.ID;
```

For formats that support integration with the schema registry, the key schema will be read and 
registered with the Schema Registry as needed, following the same pattern as the value schema in 

> the current product.

Reviewer 2: This is also an interesting point with regard to _when_ the format changes happens -- internal topics might register different stuff in the SR -- if we let the optimizer decide, and we change the optimization rules in the future, it raises schema compatibility questions.

Author: Nope, no concerns at all.   See the compatibility section.

Author: Again, see https://github.com/confluentinc/ksql/blob/dcd28ff1b26705625a477db4e3e81a1891b184aa/design-proposals/klip-33-key-format.md#internal-topics

### New server config

In addition to new `CREATE` properties, the user will also be able to set default key and value 
formats using system configuration. 

> `KEY_FORMAT` will not be a required property _if_ `ksql.persistence.default.format.key` is set. 

Reviewer: I'm wondering if we should default the default to `KAFKA` - it's not unlikely that most users will have `KAFKA` formatted keys and this will make it a better OOTB experience

Author: Already suggested, see below.

`VALUE_FORMAT` is currently a required property, but will no longer be required _if_ 
`ksql.persistence.default.format.value` is set.

Somewhat unrelated, but small, we propose also changing the `KAFKA_TOPIC` property from _required_ to 
_optional_ in `CREATE STREAM` and `CREATE TABLE` statements. If not supplied, it will default to the 
uppercase name of the stream or table.  This matches the behaviour of the property in 
`CREATE STREAM AS SELECT` and `CREATE TABLE AS SELECT` statements, where it is already optional.

> To additional configurations will be added to control topic names:

Author: Originally, this was a tiny little additional piece of work. But its growing, so will pull into a separate KLIP.

> To additional configurations will be added to control topic names:

Author: KLIP for making `WITH` clause optional: https://github.com/confluentinc/ksql/pull/6065

* `ksql.persistence.topic.name.case`: if `lower`, defaults ksql to using lowercase topics names
  where topic names are not explicitly provided. This includes matching a `CREATE TABLE` or 
  `CREATE STREAM` statement to an existing topic. Default: `upper`.
* `ksql.persistence.topic.name.case.insensitive`: if `true`, matching a `CREATE TABLE` or 
  `CREATE STREAM` statement to an existing topic will be case insensitive. Multiple matches will 
  result in an error. Default: `false`.

With these changes there will no longer be any required `CREATE` properties. This means the `WITH` 

> clause for `CREATE` statements for streams and tables would be optional. We propose this makes the 

Reviewer: I suppose we still need `partitions` to indicate that the topic should be created, unless we want to re-intorudce a default there

Author: I had thought the same.  Yes, we could include a default partition count.  I don't think adding one would cause any problems with the current 'use existing' vs 'create new' logic.

Author: I'll add this to the KLIP I'm raising for making the `WITH` clause optional

Author: KLIP for making `WITH` clause optional: https://github.com/confluentinc/ksql/pull/6065

```sql
SET 'ksql.persistence.default.format.key'='Avro';
SET 'ksql.persistence.default.format.value'='Avro';

CREATE TABLE USERS (
  ID BIGINT PRIMARY KEY,
  NAME STRING
);

CREATE STREAM BIDS (
  ITEM_ID BIGINT KEY,
  USER_ID BIGINT,
  AMMOUNT INT
);
```

### Implementation

The system already serializes the key format of source, intermediate and sink topics as part of the
query plan, meaning it should be fairly easily to plug in new formats. 

Validation will be added to ensure only supported key formats are set, and that key column data types
are supported by key formats.

Most existing functionality should _just work_, as the key format only comes into play during 
(de)serialization, (obviously). The only area where additional work is expected is joins.

Joins require the binary key of both sides of the join to match and both sides to be delivered to 
the same ksqlDB node.  The former normally ensuring the latter, unless a custom partitioning 
strategy has been used.

The introduction of additional key formats means that while the deserialized key from both sides of 
a join may match, the serialized binary data may differ if the key serialization format is different.

> To accommodate this, ksqlDB will automagically repartition the right side of a join to match the key

Reviewer 2: Why the right side?

Author: Because it's better than the wrong side, obviously!

Unlike a traditional system, we can't easily _know_ which side is the cheapest to repartition.  When queries are started the topics in question could be empty.  Even if they aren't empty, we can't really read anything into the size of the topic. A topic may be huge, while the other side tiny, but that's because the huge topic contains a pretty static table, and the other side is about to have a fire hose turned on. We can't even assume a large change log means a large table, it could just be bad/no compaction.

Without an optimiser to choose which side to repartition, I figured we just have to pick one.  And it felt more likely that the right would be the cheaper, on average, to repartition, mainly because of stream-table joins, where the table doesn't trigger joins and will often be updated less frequently than the right.

The user is always free to explicitly repartition themselves if they don't like our choice.

Of course, this is just a gut feeling. So feel free to plow in and brake all my assumptions ;)

Author: I've extended the detail about how we will pick the formats during joins, though the actual algo is still not defined: https://github.com/confluentinc/ksql/blob/dcd28ff1b26705625a477db4e3e81a1891b184aa/design-proposals/klip-33-key-format.md#implementation

format of the left side.  Such repartitioning is possible and safe, even for tables, because the 
logical key of the data will not have changed, only the serialization format. This ensures the 

> ordering of updates to a specific key are maintained across the repartition.

Reviewer 2: It may still introduce out-of-order data across different keys and thus time-tracking ("stream-time"), and grace-period and retention-time might be affected. (But it might be the same issue if we pick the left-input for repartitioning?).

However, for a stream-table join, it might be simpler to repartition the stream-side? Most likely, the join won't happen on the original key and the stream must be repartioned anyway (and we can just change the format on-the-fly)?

Author: > It may still introduce out-of-order data across different keys and thus time-tracking ("stream-time"), and grace-period and retention-time might be affected. (But it might be the same issue if we pick the left-input for repartitioning?).

Sure, just like any other implicit repartitioning we do.  Not a lot we can do about it, other than call it out in the docs.

> However, for a stream-table join, it might be simpler to repartition the stream-side? Most likely, the join won't happen on the original key and the stream must be repartitioned anyway (and we can just change the format on-the-fly)?

Good point about leveraging any existing repartition to also change the format. (I'll add this to the KLIP) If there is no existing repartition, which side to choose is fairly arbitrary, however, I've outlined why I think it should be the right in my [comment](https://github.com/confluentinc/ksql/pull/6017#discussion_r473983198) about.

Reviewer 3: i think the idea that, in a stream-table join, it's the stream that gets repartitioned (assuming any repartitioning needed at all of course) makes intuitive sense. If you think about it, you _should_ only join to the table on it's key anyway - unless you really understand the implications at least. 
Obviously table-table joins that require a re-partition should be delegated to the kstreams FK-join feature to handle. (I know that's not implemented into ksqlDB yet but it hopefully won't be too far in the future - seems sub-optimal to build some strange short-term workaround for it?)
That leaves stream-stream joins, for which I agree that some arbitrary rule is required, hopefully one that's easily remembered like "left side always gets repartitioned" :-)

    * schema registry integration
    * making KAFKA_TOPIC property optional
    * Automatic repartitioning of streams and tables for joins where key formats do not match: such
      joins will result in an error initially.
  
    Included in this milestone:
    
    * Addition of a new optional `KEY_FORMAT` property in the `WITH` clause, to set the key format.
    * Addition of a new optional `FORMAT` property in the `WITH` clause, to set both the key & value formats.
    * Addition of new server configuration to provide defaults for key and value formats.
    * Removal of requirement for `VALUE_FORMAT` property in the `WITH` clause of CT/CS statements, where
      the server configuration provides a default.
    * Support for additional key column data types, as JSON supports them:
        * `DECIMAL`
        * `BOOLEAN`
    * Full support of these key formats for all supported SQL syntax  
1. **Auto-repartitioning on key format mismatch**. Adds support for automatic repartitioning of 
   streams and tables for joins where key formats do not match.
1. **Schema Registry support**: Adds support for reading and writing schemas to and from the schema
   registry.

> 1. **JSON_SR support** Adds support for the `AVRO` key format, inc. schema registry integration.

Reviewer 4: ```suggestion
1. **JSON_SR support** Adds support for the `JSON_SR` key format, inc. schema registry integration.
```

1. **Avro support** Adds support for the `AVRO` key format, inc. schema registry integration.
1. **Delimited support**: Adds support for the `DELIMITED` key format.
1. **Optional KAFKA_TOPIC property**: makes the `KAFKA_TOPIC` property optional everywhere and add
   configuration to control case and case sensitivity of topic names.
1. **Blog post**: write a blog post about the new features. (Potentially more than once if 
  work span multiple releases).
   
LOE TBD once scope and design confirmed.

## Documentation Updates

New server config and new `CREATE` properties will be added to main docs site.

There are no incompatible changes within the proposal, so no demos and examples _must_ change.
However, it probably pays to update some to highlight the new features. We propose updating the 
Kafka micro site examples to leverage the new functionality, as these have automated testing.  

> It may be worth changing the ksqlDB quickstart too - TBD, as this will require extending DataGen 

Reviewer 4: Which quickstart does this refer to? Neither the on-prem nor cloud quickstart uses ksql-datagen.

Author: Good to know!

Author: Updated.

to support other key formats. Something we may want in scope anyway - or should be end-of-life 

> DataGen in favour of the datagen connector?

Reviewer: +1

Reviewer 3: Please don't eol the datagen tool! 
It's far too useful for anyone who wants to test or demo something quickly. And before someone says "you can just run an embedded datagen connector though so what's the difference?" :-),  allow me to opine that this requires a steeper learning curve and advanced troubleshooting skills to get working right, which I think should be avoided where possible. Just an opinion of course ;)
I'd also say that the connector should be refactored (as i recall suggesting loudly when it was originally forked off from here) so that it embeds some re-usable portion of the datagen tool (which likely requires a small refactor of datagen itself too, to facilitate, to be fair) rather than being a copy/paste that now proceeds on its own life journey and inevitable divergence.

Author: @blueedgenick are you putting a case forward that DataGen should be enhanced to support Avro / Json keys as part of this work, i.e. it should be in-scaope?

## Compatibility Implications

As mentioned above, existing query plays already include key formats for all topics. So existing
queries will continue to work.

Without `ksql.persistence.default.format.key` set to `KAFKA` existing queries in the form:

```sql
CREATE TABLE USERS (
    ID BIGINT PRIMARY KEY,
    NAME STRING
  ) WITH (
    KAFKA_TOPIC='USERS',
    VALUE_FORMAT='JSON'
  );
```

...will start failing, as they do not specify the `KEY_FORMAT`. We therefore propose the server 

> config shipped with ccloud and on-prem releases has `ksql.persistence.default.format.key` set to 

Author: The former.  

If the `KSqlConfig` doesn't define a default, i.e. the default is in the server config file we ship, then we don't need `NONE`.  If it does, then `NONE` may be needed. TBD.

> `KAFKA`.

Reviewer: +1 as above, why not have this be the in-code default?

Author: I see two main ways of implementing this:
 - default in the `KsqlConfig`
 - defaults in the server properties files.

The former has the pro of handling both on-prem and ccloud, but the con that it might not be possible to unset the property, i.e. to add config to say 'I don't want a default'.  Not sure that's an issue though.

such solutions tend to increase the complexity of the system, and generally hurt performance.

To open ksqlDB up to new problems spaces and to drive adoption, ksqlDB should support other key formats. 

## What is in scope

 * Addition of a new optional `KEY_FORMAT` property in the `WITH` clause, to set the key format.
 * Addition of a new optional `FORMAT` property in the `WITH` clause, to set both the key & value formats.
 * Addition of new server configuration to provide defaults for key format.
 * Support for additional key column data types, where the key format supports it:
    * `DECIMAL`
    * `BOOLEAN`
 * Support of the following key formats:
    * `KAFKA`: the current key format.
    * `DELIMITED`: single key columns as a single string value.
    * `JSON` / `JSON_SR`: single key column as an anonymous value, i.e. not within a JSON object.
    * `AVRO`: single key column as an anonymous value, i.e. not within an Avro record.
    * `NULL`: special format indicating no data, or ignored data, e.g. a key-less stream.
  * Full support of these key formats for all supported SQL syntax.
  * Automatic repartitioning of streams and tables for joins where key formats do not match.

>   * Support for reading & writing key schemas to & from the schema registry. 

Author: todo: Enhancing QTT, testing too ... and new SQL testing tool?

Reviewer: If I've been coding it as I intended, YATT should get this KLIP for free 😂

  FROM BIDS 
   JOIN USERS ON BIDS.USER_ID = USERS.ID;
```

For formats that support integration with the schema registry, the key schema will be read and 
registered with the Schema Registry as needed, following the same pattern as the value schema in 
the current product. 

If a `CREATE TABLE` or `CREATE STREAM` statement does not include a `KEY_FORMAT` property, the 
key format is picked up from the `ksql.persistence.default.format.key` system configuration. If this
is not set, then an error is returned.  Note: The server config will have this set to `KAFKA` to 
maintain backwards compatibility with current system by default.

### Implementation

The system already serializes the key format of source, intermediate and sink topics as part of the
query plan, meaning it should be fairly easily to plug in new formats. 

Validation will be added to ensure only supported key formats are set, and that key column data types
are supported by key formats.

Most existing functionality should _just work_, as the key format only comes into play during 
(de)serialization, (obviously). The only area where additional work is expected are joins, single
key wrapping and key-less streams.

> #### Joins

Reviewer 5: Sorry I am late to the party. 
What happens in this scenario: ```CREATE TABLE test as SELECT * from test1 join test2``` where the `test1` join key has format `f1` and `test2` join key has format` f2` . The output key format is inherited from the left-most source hence will be `f1`? If the algorithm chooses that the cheapest partitioning strategy is to partition by `f2`, does this mean that we need an extra partitioning at the end for the sink topic to be in `f1` format?

If yes, then this should be part of the decision choosing which side to repartition, no?
So the heuristics for the algorithm should take into account:
1) What is the output format? 
2) What keys are the sources partitioned by? 

And the definition of cheapest should take into account the cost of partitioning at the end to match the final output format as well as the cost of re-partitioning the inputs.

Reviewer 5: In addition to this, will this be supported 
``` 
CREATE TABLE TEST 
  WITH (
    KEY_FORMAT='F3'
  ) AS
    SELECT * FROM TEST1 JOIN TEST2;
```
If yes, then the output format can differ completely from the input formats which will definitely necessitate a final repartition for the sink topic.

Reviewer 2: > which will definitely necessitate a final repartition for the sink topic.

Well, yes, but this is not really an issue as the "repartitioning" is for free: we write the data into the sink topic anyway and the costs is the same with or without changing the partitioning.

Reviewer 5: How is the cost the same? If the partitioning changes, isn't there network overhead and data shuffling involved?

Reviewer 2: There is not more network overhead involved. Every message is written once.

Note, in general when we talk about repartitioning overhead, the issue is that we need to write the data and read it back to shuffle records. (And with no repartitioning, we don't neither write nor read to a topic, ie, the "overhead/cost" of repartitioning is one write and one read operation per record.) -- However, when we write to a sink anyway, we don't have more write operations if we change the partitioning.

Maybe, the only "overhead" we might have is that the producer may need to open more network connections as it may need to write data into different partitions that might be hosted on different brokers. However, this seems negligible.

# KLIP 33 - Key format

**Author**: @big-andy-coates | 
**Release Target**: 0.13 | 
**Status**: In Discussion | 
**Discussion**: https://github.com/confluentinc/ksql/pull/6017

**tl;dr:** ksqlDB currently only supports keys compatible with the `KAFKA` format. This limits the
           data ksqlDB can work with. Extending the set of key formats ksqlDB supports immediately
           opens up the use of ksqlDB with previously incompatible datasets.
           
## Motivation and background

Data stored in Kafka has a key and a value. These can be serialized using different formats, but
generally use a common format. ksqlDB supports multiple _value_ [data formats][1], but requires the 
key data format to be the `KAFKA`.  

This limitation is particularly problematic for tables. ksqlDB is unable to access changelog topics
in Kafka that have non-`KAFKA` formatted keys. As the key of the Kafka record is the `PRIMARY KEY` 
of the table, it is essential that the key can be read if the changelog is to be materialised into a 
table. When changelog topics have non-`KAFKA` key formats, the limitation precludes ksqlDB as a solution.

> The limitation is less damaging for streams. However, it is still the case that the user looses the

Reviewer 6: ```suggestion
The limitation is less damaging for streams. However, it is still the case that the user loses the
```

## What is not in scope

 * Support for multiple key columns: this will come later.
 * Support for single key columns _wrapped_ in an envelope of some kind: this will come later.
 * Support for complex key column data types, i.e. array, struct and map: this will come later. 
 * Support for `PROTOBUF` keys, as this requires support for wrapped keys: this will come later.
 * Enhancing DataGen to support non-KAFKA keys.
 * Key schema evolution. (See [key schema evolution](#key-schema-evolution)) in the compatibility 
   section.
 * Support for right-outer joins. This may be covered in a future KLIP.

## Value/Return

We know from customers and community members that there are a lot of people that have data with 
non-`KAFKA` formatted keys. This is the first step to unlocking that data and use-cases.

With support for `AVRO` and `JSON` key formats there are a lot of existing use-cases that suddenly
no longer require pre-processing, or tricky Connect SMTs configured, and there are new use-cases,
which ksqlDB was previously unsuitable for, as documented in the motivation section, which can 
now be handled.

> ## Public APIS

Reviewer 6: ```suggestion
## Public APIs
```

### CREATE properties

The following new properties will be accepted in the `WITH` clause of `CREATE` statements for streams
and tables.

* `KEY_FORMAT`: sets the key format, works long the same lines as the existing `VALUE_FORMAT`.
* `FORMAT`: sets both the key and value format with a single property.

`KEY_FORMAT` will not be a required property _if_ `ksql.persistence.default.format.key` is set.

Providing `FORMAT` will set both the key and value formats. Providing `FORMAT` along with either 
`KEY_FORMAT` or `VALUE_FORMAT` will result in an error.

### Server configs

The following new configuration options will be added. These configurations can be set globally, 
within the application property file, or locally, via the `SET` command.

* `ksql.persistence.default.format.key`: the default key format.

## Design

> The new `KEY_FORMAT` or `FORMAT` property will be supported where ever the current `VALUE_FORMAT` is

Reviewer 6: ```suggestion
The new `KEY_FORMAT` or `FORMAT` property will be supported wherever the current `VALUE_FORMAT` is
```

supported. Namely:

>  * In `CRREATE STREAM` and `CREATE TABLE` statements.

Reviewer 6: ```suggestion
 * In `CREATE STREAM` and `CREATE TABLE` statements.
```

 * In `CREATE STREAM AS SELECT` and `CREATE TABLE AS SELECT` statements.
 
The key format will follow the same inheritance rules as the current value format. Namely: any 
derived stream will inherit the format of its leftmost source, unless the format is explicitly set

> in the with clause.

Reviewer 6: ```suggestion
in the `WITH` clause.
```

are supported by key formats.

Most existing functionality should _just work_, as the key format only comes into play during 
(de)serialization, (obviously). The only area where additional work is expected are joins and key-less 
streams.

#### Joins

Joins require the binary key of both sides of the join to match and both sides to be delivered to 
the same ksqlDB node.  The former normally ensuring the latter, unless a custom partitioning 
strategy has been used.

The introduction of additional key formats means that while the deserialized key from both sides of 
a join may match, the serialized binary data may differ if the key serialization format is different.
To accommodate this, ksqlDB will automagically repartition one side of a join to match the key
format of the other.

Many joins require an implicit repartition of one or both sides to facilitate the join. In such 
situations the change of key format can be performed in the same repartitioned step, avoiding any
additional re-partitions. This means that joining sources with different key formats will only 
require an implicit repartition to converge the key formats _if_ neither side is already being 
repartitioned.

> Where one side must be repartitioned to correct the key format, choosing which side to reparation 

Reviewer 6: ```suggestion
Where one side must be repartitioned to correct the key format, choosing which side to repartition 
```

the query, with the source on the _right_ being repartitioned.

A benefit of making the choice order-based is that, once the rule is learned, users can predicate 
and control which side is re-partitioned in some situations, i.e. stream-stream and table-table joins.

Note: allowing users to freely switch left and right sources to control which side is repartitioned
will work for all but left-outer joins. To support switching left-outer joins ksqlDB would need to
support a right-outer join. The addition of this is deemed out of scope.

Repartitioning the right side was chosen over the left, as it will mean stream-table joins will 
repartition the table, which we propose will _generally_ see a lower throughput of updates to the 
stream side. 

Such repartitioning is possible and safe... ish, even for tables, because the logical key of the 
data will not have changed, only the serialization format. This ensures the ordering of updates to 
a specific key are maintained across the repartition. Of course, the repartitioning would introduce 
cross-key out-of-order data, as the records are shuffled across partitions. That is to say that 

> even if the source partitions where correctly ordered by time, the re-partitioned partitions would 

Reviewer 6: ```suggestion
even if the source partitions were correctly ordered by time, the re-partitioned partitions would 
```

see out-of-order records, though per-key ordering would be maintained. Thus time-tracking 
("stream-time"), grace-period and retention-time might be affected. However, this  phenomenon 
already exists, and is deemed acceptable, for other implicit re-partitions.

> 

Reviewer 4: Add note clarifying that users can always repartition topics themselves before the join, in order to have full control over which sources are repartitioned (and that choosing the same sources ksqlDB would repartition and performing the repartitions upfront is equivalent from a resource-usage standpoint)? Or if not here, at least in the docs section so we don't forget to add the note later.

below for more info / background.

#### Key-less streams

A new `NONE` format will be introduced to allow users to provide a `KEY_FORMAT` that informs ksqlDB
to ignore the key. This format will be rejected as a `VALUE_FORMAT` for now, as ksqlDB does not yet
support value-less streams and tables. See [Schema Inference](#schema-inference) below for more 
info / background.

This format is predominately being added to allow users to declare key-less streams when the new 
`ksql.persistence.default.format.key` system configuration is set to a format that supports schema
inference, i.e. loading the schema from the schema registry. If a user were not to explicitly set
the key format to `NONE` and attempt to create a stream, ksqlDB would attempt to read the key schema
from the schema registry, and report an error if the schema did not exist. The `NONE` format will 
allow users to override the default key format and explicitly inform ksqlDB to ignore the key:

```sql
SET 'ksql.persistence.default.format.key'='AVRO';

-- Only the value columns of CLICKS will be loaded from the schema registry.
CREATE STREAM CLICKS 
 WITH (

>    key_format='NONE',  -- Informs kdqlDB to ignore the key

Reviewer 4: ```suggestion
   key_format='NONE',  -- Informs ksqlDB to ignore the key
```

1. **Schema Registry support**: Adds support for reading and writing schemas to and from the schema
   registry.
1. **JSON_SR support (1 week)** Adds support for the `JSON_SR` key format, inc. schema registry integration.
1. **Avro support (1 week)** Adds support for the `AVRO` key format, inc. schema registry integration.
1. **Delimited support (1 week)**: Adds support for the `DELIMITED` key format.
1. **Auto-repartitioning on key format mismatch (1.5 weeks)**. Adds support for automatic repartitioning of 
   streams and tables for joins where key formats do not match.
1. **Blog post (1 week)**: write a blog post about the new features. Likely one post for everything _but_ 
   auto-repartitioning, and a second to cover this.
   
## Documentation Updates

New server config and new `CREATE` properties will be added to main docs site.

There are no incompatible changes within the proposal, so no demos and examples _must_ change.
However, it probably pays to update some to highlight the new features. We propose updating the 
Kafka micro site examples to leverage the new functionality, as these have automated testing.  

> It may be worth changing the ksqlDB quickstart too. 

Reviewer 6: :+1:
